src/table_properties.rs

```diff
@@ -0,0 +1,246 @@
+use std::ffi::{c_char, c_void, CStr, CString};
+use std::marker::PhantomData;
+use std::mem;
+use std::slice;
+
+use libc::{c_int, size_t};
+
+use crate::{ffi, Options};
+
+/// Extension trait for [`Options`] to register table properties collectors
+pub trait TablePropertiesExt {
+    fn add_table_properties_collector_factory<F>(&mut self, factory: F)
+    where
+        F: TablePropertiesCollectorFactory + Send + 'static;
+}
+
+impl TablePropertiesExt for Options {
+    fn add_table_properties_collector_factory<F>(&mut self, factory: F)
+    where
+        F: TablePropertiesCollectorFactory + Send + 'static,
+    {
+        unsafe {
+            let factory_ptr = Box::into_raw(Box::new(factory)) as *mut c_void;
+
+            // Takes ownership of the collector factory; the Rust callback wrapper will be
+            // dropped via the destructor callback
+            ffi::rocksdb_options_add_table_properties_collector_factory(
+                self.inner,
+                factory_ptr,
+                Some(TablePropertiesCollectorFactoryCallback::<F>::destructor),
+                Some(TablePropertiesCollectorFactoryCallback::<F>::name),
+                Some(TablePropertiesCollectorFactoryCallback::<F>::create_collector),
+            );
+        }
+    }
+}
+
+#[repr(C)]
+pub enum EntryType {
+    EntryPut,
+    EntryDelete,
+    EntrySingleDelete,
+    EntryMerge,
+    EntryRangeDeletion,
+    EntryBlobIndex,
+    EntryDeleteWithTimestamp,
+    EntryWideColumnEntity,
+    EntryTimedPut,
+    EntryOther,
+}
+
+pub struct TablePropertiesCollectorContext {
+    pub column_family_id: u32,
+    pub level_at_creation: i32,
+    pub num_levels: i32,
+    pub last_level_inclusive_max_seqno_threshold: u64,
+}
+
+/// Table properties collector factory trait
+pub trait TablePropertiesCollectorFactory {
+    type Collector: TablePropertiesCollector;
+
+    /// Create a new table properties collector
+    fn create(&mut self, context: TablePropertiesCollectorContext) -> Self::Collector;
+
+    /// Name of the collector factory to use for logging
+    fn name(&self) -> &CStr;
+}
+
+/// Table properties collector trait
+pub trait TablePropertiesCollector {
+    /// Called when a new key/value pair is added to the table
+    fn add_user_key(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        entry_type: EntryType,
+        seq: u64,
+        file_size: u64,
+    ) -> Result<(), ()>;
```

Comment:
I was hoping to use the crate's own `rocksdb::Error` here but construction is meant to be private to the crate. It seems way out of scope to model RocksDB's `Status` as a Rust error type - particularly given how RocksDB's usage of these methods amounts to OK/not-OK. What's a good course of action here? I want `add_user_key` to be consistent with `finish` and return `Result`, but it also feels awkward returning `Result<_, ()>`.